### PR TITLE
[TECHDEBT][PERSISTENCE] Add: connection pooling to persistence layer db connection

### PR DESCRIPTION
**Pull request by Prajjawalk**
## Description

The current database connection is established via pgx.Connect() method which is not concurrency safe, so updated the database connection to connect via pgxpool module which establishes connection pooling. Updated configs to apply more connection configurations: max-connections, min-connections, max-connection lifetime, max-connection idle-time and health check intervals.  

## Issue

Fixes #NA. It was a TPDP w/o an issue

## Type of change

Please mark the relevant option(s):

- [X] New feature, functionality or library
- [ ] Bug fix
- [X] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Upgrade `github.com/jackc/pgx/v4` -> `github.com/jackc/pgx/v5`
- Modified `persistence/db.go` -> `connectToDatabase()` function to use pgxpool for connecting to db
- Modified `persistence_config.proto`, `config1.json`, `config2.json`, `config3.json` and `config4.json` to include connection pool configurations

## Testing

- [X] `make develop_test`
- [X] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have tested my changes using the available tooling
- [X] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: persistence/db.go
Comment (Olshansk) [inline persistence/db.go:71]: Is it safe to remove this TECHDEBT line item?

File: shared/modules/types.go
Comment (Olshansk) [inline shared/modules/types.go:51]:
```suggestion
	GetMaxConnsCount() uint32
```

File: shared/modules/types.go
Comment (Olshansk) [inline shared/modules/types.go:52]: Why do we need a min?

File: shared/modules/types.go
Comment (Olshansk) [inline shared/modules/types.go:53]: For these three, can you add a postfix explaining the time parameter? The code speaks for itself then

E.g.

`s/GetMaxConnLifetime/GetMaxConnSec`
`s/GetMaxIdleTime/GetMaxIdleSec`

File: persistence/module.go
Comment (Olshansk) [inline persistence/module.go:186]: If we're going to be passing in the whole config, then there's no need for a second parameter.

**Review by Olshansk — CHANGES_REQUESTED** [review]
Thank you for your contribution! I'm really curious how you found this specific task to work on?

Left a few comments throughout, primarily around the naming so the measurement units are self-explanatory.

File: persistence/db.go
Comment (Prajjawalk) [reply persistence/db.go:71]: Hi @Olshansk, Thanks for reviewing. I was just going through the code and found this Techdebt item which was not resolved so worked on that. I will go through the review and will make changes accordingly, after that we can remove this line.

File: shared/modules/types.go
Comment (Prajjawalk) [reply shared/modules/types.go:52]: Min is by default 0 but added it in case we may need it in the future.

File: shared/modules/types.go
Comment (Prajjawalk) [reply shared/modules/types.go:53]: Yep sure, will do

File: shared/modules/types.go
Comment (Olshansk) [inline shared/modules/types.go:52]: Thoughts on making this unsigned?

EDIT: My guess is that you made them signed to correspond with https://pkg.go.dev/github.com/jackc/pgx/v4/pgxpool?

File: shared/modules/types.go
Comment (Olshansk) [inline shared/modules/types.go:55]: I looked into: https://pkg.go.dev/github.com/jackc/pgx/v4/pgxpool

And found that it uses found time.Duration: https://pkg.go.dev/time#Duration

It seems that we can easily parse things like `3s` or `1h` in golang: https://www.geeksforgeeks.org/time-parseduration-function-in-golang-with-examples/

Thoughts on removing the `Hour`, `Minute`, etc.., but making these strings so it could be configured to seconds, minutes, hours, etc in the config level.

This will also be similar to how tendermint designed things: github.com/pokt-network/pocket

Screenshot 2022-12-21 at 12 24 35 PM

**Review by Olshansk — CHANGES_REQUESTED** [review]
Left another couple comments. Let me know what you think.

Make sure to also use `LocalNet` to test and update the `CHANGELOG`.

File: shared/modules/types.go
Comment (Prajjawalk) [reply shared/modules/types.go:52]: Yep, I had to make the connection count type int32 to correspond to pgxpool connection count type

File: shared/modules/types.go
Comment (Prajjawalk) [reply shared/modules/types.go:55]: Makes sense, will make changes accordingly

File: persistence/db.go
Comment (Olshansk) [reply persistence/db.go]: 💯

File: persistence/db.go
Comment (Olshansk) [inline persistence/db.go:78]: Local var should start with a lowercase letter. Ditto below

File: persistence/proto/persistence_config.proto
Comment (Olshansk) [inline persistence/proto/persistence_config.proto:12]: Consider adding a comment above the proto saying "See https://pkg.go.dev/time#ParseDuration for expected string time parameters"

File: persistence/db.go
Comment (Olshansk) [inline persistence/db.go:82]: Let's return an error if parsing fails/

I understand the logic of "a quite success" of using the defaults, but it could lead to a bug where someone incorrectly configures a time parameter and spends hours debugging why it wasn't applied.

**Review by Olshansk — CHANGES_REQUESTED** [review]
Left a couple of comments.

Also, can you rebase/merge this on top of #403 (or wait until it is merged to main)? We are doing an overhaul of how the configs are managed and I'm expecting some merge conflicts.

Apologize for the extra work!

File: persistence/db.go
Comment (Prajjawalk) [reply persistence/db.go:78]: Yep, shall I raise PR on top of "396-consolidate-config" branch?

Comment (Olshansk) [reply persistence/db.go]: Yup

**Comment by Olshansk** [comment]
@Prajjawalk Could you close #424, merge this one with main (resolving any conflicts) and reupload it for review after testing?

I believe it should be good to merge then.

File: persistence/proto/persistence_config.proto
Comment (Olshansk) [reply persistence/proto/persistence_config.proto]: @Prajjawalk Wdyt?

File: persistence/db.go
Comment (Olshansk) [inline persistence/db.go:82]: @Prajjawalk Do you think just logging it (rather than returning an error) is more appropriate?

I was thinking that if we return `nil, err`, it'll inform the operator their configs are misconfigured and fix it. Not a lot of people look at logs :)

File: runtime/configs/proto/persistence_config.proto
Comment (Olshansk) [inline runtime/configs/proto/persistence_config.proto:15]:
```suggestion
  string max_conn_lifetime = 8; # See pkg.go.dev/time#ParseDuration for reference
```

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: persistence/db.go
Comment (Prajjawalk) [reply persistence/db.go:82]: If we return an error instead of logging then it may break the execution flow throwing a runtime error, I thought if configs are misconfigured then the error may not be that critical to break the flow as we can resort to default while informing user through logs, what do you think @Olshansk ?

File: persistence/db.go
Comment (Olshansk) [inline persistence/db.go:108]: Can you create a local var for `schema := cfg.GetNodeSchema()` and reuse it in the three places below?

File: persistence/db.go
Comment (Olshansk) [reply persistence/db.go]: I don't think there's a right answer here as I can definitely argue both sides of it. I'd say it's probably a function of what were the worse situation a developer has experienced in the past:

1. Changing configs and spending a lot of time figuring out why they weren't being applied.
2. Misconfiguring configs and having everything crash.

There's a case to be made for each :)

**Review by Olshansk — CHANGES_REQUESTED** [review]
One last comment and then it's good to go. Thanks for bearing with the reviews and the awesome change :)

File: persistence/db.go
Comment (Prajjawalk) [reply persistence/db.go]: yep

**Review by Olshansk — APPROVED** [review]
LGTM! Thanks for the high quality work @Prajjawalk!

Can you rebase on top of master so we can :shipit:

**Comment by Olshansk** [comment]
@Prajjawalk Could you please:

- [ ] See the failing CI issue and tend to it
- [ ] Confirm that you've tested this manually with Localnet to confirm everything works. 👍 ?

**Comment by Prajjawalk** [comment]
> @Prajjawalk Could you please:
> 
> * [ ]  See the failing CI issue and tend to it
> * [ ]  Confirm that you've tested this manually with Localnet to confirm everything works. 👍 ?



> @Prajjawalk Could you please:
> 
> * [X]  See the failing CI issue and tend to it
> * [X]  Confirm that you've tested this manually with Localnet to confirm everything works. 👍 ?

**Comment by Olshansk** [comment]
I fixed the dates one more time and also verified that "everything on my machine". Thank you for the contribution and sorry for the delay.

Let me know if you're interested in helping out more :)

**Pull request merged**
